File: Dart/heapsort.dart
//Heap sort function
List<int> heapSort(MinHeap minHeap) {
  List<int> sortedList = [];
  
  while (minHeap.heap.isNotEmpty) {
    int minElement = minHeap.peek();
    sortedList.add(minElement);
    minHeap.remove();
  }

  return sortedList;
}
// construction of heap
class MinHeap {
  List<int> heap = [];

  MinHeap(List<int> list) {
    buildHeap(list);
  }

  void buildHeap(List<int> list) {
    heap = list;
    for (int i = parent(heap.length - 1); i >= 0; i--) {
      shiftDown(i);
    }
  }

  void shiftDown(int current) {
    int end = heap.length - 1;
    int left = leftChild(current);

    while (left <= end) {
      int right = rightChild(current);
      int idxToShift;

      if (right <= end && heap[right] < heap[left]) {
        idxToShift = right;
      } else {
        idxToShift = left;
      }

      if (heap[current] > heap[idxToShift]) {
        swap(heap, current, idxToShift);
        current = idxToShift;
        left = leftChild(current);
      } else {
        return;
      }
    }
  }

  void shiftUp(int current) {
    int parents = parent(current);
    while (current > 0 && heap[parents] > heap[current]) {
      swap(heap, current, parents);
      current = parents;
      parents = parent(current);
    }
  }

  int parent(int i) {
    return (i - 1) ~/ 2;
  }

  int leftChild(int i) {
    return 2 * i + 1;
  }

  int rightChild(int i) {
    return 2 * i + 2;
  }

  void swap(List<int> list, int i, int j) {
    int temp = list[i];
    list[i] = list[j];
    list[j] = temp;
  }

  int peek() {
    return heap[0];
  }

  void display() {
    for (int i = 0; i < heap.length; i++) {
      print(heap[i]);
    }
  }

  void remove() {
    swap(heap, 0, heap.length - 1);
    heap.removeLast();
    shiftDown(0);
  }

  void insert(int data) {
    heap.add(data);
    shiftUp(heap.length - 1);
  }
}




// for example main function is provided

void main(List<String> args) {
  List<int> list = [9, 7, 8, 6, 3, 4, 5, 1];
  MinHeap minHeap = MinHeap(list);
  print("Min Heap:");
  minHeap.display();
  List<int> sortedList = heapSort(minHeap);
  print("\nSorted Array:");
  for (int i = 0; i < sortedList.length; i++) {
    print(sortedList[i]);
  }
}

// Solution is commented

/* 

Min Heap:
1
3
4
6
9
8
5
7

Sorted Array:
1
3
4
5
6
7
8
9

*/

File: Python/Heap Sort.py

#* heap sort in ascending order 

def heapify(arr,len,i):
    par=i
    left=2*i +1
    right=2*i +2                             #* descending order
                                            
    if left<len and arr[par]<arr[left]:    #* if left<len and arr[par]<arr[left]:
        par=left
        
    if right<len and arr[par]<arr[right]:  #* if right<len and arr[par]>arr[right]:

        par=right
        
    if par != i:
        arr[i],arr[par]=arr[par],arr[i]
        heapify(arr,len,par)
        
def heap_sort(arr):
    
    l=len(arr)
    for i in range(l//2 -1, -1, -1):
        heapify(arr,l,i)
        
    for j in range(l-1,0,-1):
        arr[0],arr[j]=arr[j],arr[0]
        heapify(arr,j,0)
        
    return arr
a=[10,30,50,28,3,5,53]
print(heap_sort(a))